### DISCUSSION TITLE
Arguments given to spawn/spawnSync on Windows breaks on spaces within quoted arguments

### DISCUSSION THREAD

**Discussion by alshdavid** [nodejs/help#4941] [Q&A]
### Version

16.15.0

### Platform

Windows 11

### What steps will reproduce the bug?

I am trying to write automation scripts (like build scripts) in JavaScript so they run across all platforms supported by Node. I have been building a wrapper around `spawn` and `spawnSync`  that guarantees execution on all environments. I cannot use any third party libraries because I don't want people to have to run `npm install` before running my scripts. I am currently having trouble with the parsing of command line arguments in Windows.

Create two files:

```javascript
// index.js
const childProcess = require('child_process')

childProcess.spawnSync('node', ['target.js', 'Hello World'], {
  shell: true,
  cwd: __dirname,
  stdio: 'inherit',
})
```
And
```javascript
// target.js
console.log(process.argv)
```

### How often does it reproduce? Is there a required condition?

Every time

### What is the expected behavior?

The terminal would have output:

```javascript
[
  'C:\\Program Files\\nodejs\\node.exe',
  'C:\\Users\\User Name\\Desktop\\foo\\target.js',
  'Hello World'
]
```

### What do you see instead?

The terminal I get the following output:

```javascript
[
  'C:\\Program Files\\nodejs\\node.exe',
  'C:\\Users\\User Name\\Desktop\\foo\\target.js',
  'Hello',
  'World'
]
```

**Comment by alshdavid** [comment]
Running `target.js` with arguments directly works as expected

![image](https://user-images.githubusercontent.com/12656294/171344626-9c3ec328-bce9-40e7-adab-3ff48253dac0.png)

 however through `spawn` it's broken. I have tried `windowsVerbatimArguments` and pointing `shell` directly to PowerShell:

![image](https://user-images.githubusercontent.com/12656294/171344681-cda6f219-de3d-47d2-a96c-e1ab082020fc.png)


For comparison, running the same operation in Golang results in the correct output:

![image](https://user-images.githubusercontent.com/12656294/171344231-d23b03d6-9b36-45dd-8b6c-5124163b88a1.png)

**Comment by bnoordhuis** [comment]
You have a bug in your code on this line:
```js
if (process.platform === 'win32' && process.arch === 'x32') {
```
Replace `x32` with `ia32`, otherwise the 64 bits version runs inside powershell while the 32 bits does not. I'll convert this to a discussion.

**Reply by alshdavid** [reply]
> You have a bug in your code on this line:
> 
> ```js
> if (process.platform === 'win32' && process.arch === 'x32') {
> ```
> 
> Replace `x32` with `ia32`, otherwise the 64 bits version runs inside powershell while the 32 bits does not. I'll convert this to a discussion.

Sorry, my wrapper isn't ideal - however when calling spawn directly without the wrapper has the same issue as seen in the images above. I will update the OP to remove the distracting wrapper:

```javascript
childProcess.spawnSync('node', ['target.js', 'Hello World'], {
  shell: true,
  cwd: __dirname,
  stdio: 'inherit',
})
```

**Reply by bnoordhuis** [reply]
Can you explain why you're passing `shell: true` when you're invoking node directly? Without `shell`, I think your example works the way you want it to.

If you do want to use `shell: true` and `windowsVerbatimArguments: true`, then you're responsible for ensuring arguments are properly quoted and escaped.

**Reply by alshdavid** [reply]
Oh ok, interesting. Why does it work when I remove `shell: true`?

Also when removing `shell: true`, I get `ENOENT` when trying to run `pnpm`, any idea why that is the case?

**Reply by bnoordhuis** [reply]
With `shell: true` you're basically invoking `cmd.exe /c "node target.js Hello World"`. Node acts as if `windowsVerbatimArguments` is true when you pass it through cmd.exe, i.e., it doesn't escape or quote.

That ENOENT is because pnpm is probably a shell script (a .bat or .cmd file.) Executing those on Windows is a property of the shell, not the operating system like it is on Unix systems.

**Reply by alshdavid** [reply]
I see. Is it possible to specify the shell as PowerShell while also automatically escaping spaces in arguments passed in?

> That ENOENT is because pnpm is probably a shell script (a .bat or .cmd file.) 

Ah yeah I see. That's correct, it's a .bat file.

**Reply by bnoordhuis** [reply]
Quoting is the default with `shell: "powershell.exe"`, it's only disabled when shelling out to cmd.exe.
